### PR TITLE
imxrt: add method to enter serial downloader

### PR DESCRIPTION
**Pull request by gerard5**
Adds `bootrom` command and ability to enter ROM serial downloader mode directly from the plo and allow download firmware using USB or boot from alternative firmware image or NOR flash.

The following command will enter NXP bootloader in USB Serial downloader mode:
```shell
(plo)% bootrom -b
```

Will boot from the 2nd image (if FCFB provided in image):
```shell
(plo)% bootrom -s 2
```

JIRA: RTOS-446

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: mimxrt1064-evk, imxrt1064-bes, mimxrt1176-evk (both PIMXRT & MIMXRT), imxrt1176-nil.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: cmds/bootloader.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <devices/devs.h>
+#include <hal/hal.h>
+#include <hal/armv7m/imxrt/bootloader.h>
+#include <lib/lib.h>
+
```
[the remark this answers is not in the document]
Comment (gerard5) [reply cmds/bootloader.c:22]: done

File: cmds/bootloader.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	if (optBootMode != 0 && optImageSel != 0) {
+		log_error("%s: select either boot mode or image number, not both\n", argv[0]);
+		print_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	optImageSel =
+		(optImageSel > 0 && optImageSel <= 4) ?
+		(optImageSel - 1) :
+		0;
```
[the remark this answers is not in the document]
Comment (gerard5) [reply cmds/bootloader.c]: done

File: cmds/bootloader.c [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		}
+	}
+
+	if ((optBootMode != 0) && (optImageSel != 0)) {
+		log_error("%s: select either boot mode or image number, not both\n", argv[0]);
+		print_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	optImageSel = ((optImageSel < 1) && (optImageSel > 4)) ? 0 : (optImageSel - 1);
```
Comment (agkaminski) [inline cmds/bootloader.c:94]: Nitpick - there's no way for `optImageSel` to be `> 4`
```suggestion
	if (optImageSel > 0) {
		--optImageSel;
	}
```

Or, more hacky way:
```
optImageSel = 1;
(...)
--optImageSel;
```

File: hal/armv7m/imxrt/bootloader.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+#ifndef _BOOTLOADER_H_
+#define _BOOTLOADER_H_
+
+
+/* Serial downloader media */
+#define BOOT_DOWNLOADER_AUTO (0x10uL << 16)
+#define BOOT_DOWNLOADER_USB  (0x11uL << 16)
+#define BOOT_DOWNLOADER_UART (0x12uL << 16)
+
+#define BOOT_IMAGE_SELECT(n) ((n)&3uL)
```
Comment (agkaminski) [inline hal/armv7m/imxrt/bootloader.h:25]: clang-format is confused sometimes
```suggestion
#define BOOT_IMAGE_SELECT(n) ((n) & 3uL)
```

Comment (gerard5) [reply hal/armv7m/imxrt/bootloader.h]: Done.

File: cmds/bootloader.c (hunk 3 above)
Comment (gerard5) [reply cmds/bootloader.c]: Done, thanks.

Comment (agkaminski) [reply cmds/bootloader.c]: I think you missed `= 1;` part though

**Review by agkaminski — APPROVED** [review]

**Pull request merged**